[webkitapipy] Document known SPI use via allowlists

emw-apple opened the pull request:
#### cbb5741c87c7827a9e28fc74a30327ded0334b88
<pre>
[webkitapipy] Document known SPI use via allowlists
<a href="https://bugs.webkit.org/show_bug.cgi?id=295295">https://bugs.webkit.org/show_bug.cgi?id=295295</a>

Reviewed by Sam Sneddon and Brianna Fan.

Introduce a mechanism for listing allowed SPI that WebKit is using. At
build time, audit-spi reads one or more project-specific allowlists,
treats those declarations as implicitly available for SPI checking, and
ensures that the list is a subset of declarations used by the project.

Add an initial set of allowlists based on what is needed to successfully
check the current supported open source build (iOS 18.5). They are
separated into &quot;legacy&quot; SPI use, which was present in WebKit when iOS 18
shipped, and SPI use added since then, which is documented for cleanup.

SPI ALLOWLISTS
==============

An allowlist is a TOML document that organizes allowed declarations into
functional categories and associates each with an *exception status*
that justifies why the SPI is allowed. For example, the entry

    [graphics.&quot;<a href="https://rdar.apple.com/147613178&quot">rdar://147613178&quot</a>;]
    symbols = [&quot;_kCAContentsFormatRGBA10XR&quot;]

allows use of the symbol _kCAContentsFormatRGBA10XR, associates it with
the &quot;graphics&quot; category, and denotes a temporary exception with
<a href="https://rdar.apple.com/147613178">rdar://147613178</a> as the cleanup bug.

More generically, an allowlist entry has three parts:

    [&lt;category&gt;.&lt;exception&gt;]
    symbols = [ ... ]
    classes = [ ... ]
    selectors = [ ... ]

The &lt;category&gt; key is an arbitrary string--its purpose is to help make
sense of why a particular SPI is being used, and to discourage
developers from simply appending to the end of the allowlist and causing
frequent merge conflicts.

The &lt;exception&gt; is a bug URL, denoting a temporary exception that will
be cleaned up when the bug is closed; or it&apos;s a fixed permanent
exception type. The symbols, classes, and selectors lists all correspond
to the different kinds of SPI that we track; the latter two are ObjC
specific.

Allowlists are required to be a subset of imported SPI. In other words,
it is an error for a declaration to be allowed but not used by any of
the input files given to audit-spi. This helps keep the list of
temporary exceptions accurate, and it&apos;s also a line of defense against
accidentally leaking internal-only SPI into public allowlists.

I chose TOML as the format because of its better human-readability
compared to JSON, and good language support. Python can read it
natively, so it does not preclude running audit-spi in offline build
environments where we cannot install PyYAML. Future work can adopt
TOML Kit to support audit-spi making edits to allowlists automatically,
for interactive development workflows.

IMPLEMENTATION CHANGES
======================

Restructure the SDKDB cache to have three main tables:
- &quot;export&quot; tracks declarations that are loaded from partial SDKDB and
  binaries, and usable as API. This merges the previous three tables.
- &quot;allow&quot; tracks declarations that are loaded from allow lists.
- &quot;import&quot; tracks declarations depended on by the input file(s) being
  analyzed. This table is ephemeral.

Instead of polling the database one declaration at a time, reimplement
the analysis as a big join query. This supports efficiently finding all
declarations which are neither allowlisted nor known to be API.
Algebraically, the join is roughly:

    ( ( import − export ) ⨯ allow ) ⨯ window

where rows that are imported but not allowed are flagged as SPI, and
rows that are allowed but not imported are flagged as needing removal.

To report results of this query, the SDKDB cache yields &quot;diagnostic&quot;
objects which are passed to the reporter for presentation.

* Source/WTF/Scripts/audit-spi-if-needed.sh: Pass --no-errors for now,
  as we are not yet ready to enforce the no-new-SPI workflow.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/_vendor/tomli: Added,
  vendored copy of &lt;<a href="https://github.com/hukkin/tomli">https://github.com/hukkin/tomli</a>&gt; (MIT License). This
  library is part of Python starting in 3.11, so we can remove it once
  we drop support for earlier versions.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py: Added.
(AllowedSPI):
(PermanentlyAllowedReason):
(AllowList): Data class that represents a single allowlist.
(AllowList.from_dict):
(AllowList.from_file): Main entrypoint.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
(TSVReporter.__init__): Recognize --errors and track &quot;suggested&quot;
  allowlists to append to when reporting unknown SPI.
(TSVReporter.emit_diagnostic): New entry point for presenting analysis
issues. Instead of querying the sdkdb cache directly, it receives
diagnostics resulting from the query in SDKDB.audit().
(TSVReporter.unused_allowed_name): Added.
(TSVReporter.finished):
(get_parser): Add a new --[no-]-errors option to control whether
potential SPI use is reported as an error.
(TSVReporter.process_report): Deleted.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(DeclarationKind): Used in the new table schema mentioned above. Rather
than have separate tables for classes/selectors/symbols, this enum is
used in a column to denote kind.
(Diagnostic): Type representing analysis issues to report. Yielded from
  audit().
(MissingName):
(UnusedAllowedName):
(SDKDB.__init__): Support marshalling DeclarationKind objects into and
  out of the database.
(SDKDB._initialize_db): Bring up new schema.
(SDKDB._initialize_temporary_schema): Ditto.
(SDKDB._add_api_report): Support adding declarations to a table other
  than the &quot;exports&quot; table, for tracking declarations that are imported or
  allowed rather than visible as API.
(SDKDB._add_objc_interface): Ditto.
(SDKDB._add_symbol): Ditto.
(SDKDB._add_objc_class): Ditto.
(SDKDB._add_objc_selector): Ditto.
(SDKDB.add_allowlist): Entry point for parsing an allowlist if needed,
  and updating the &quot;allow&quot; table. Like other input files, the allowlist
  will only be read if it&apos;s not already represented in the cache.
(SDKDB.add_for_auditing): Called to load an APIReport into the &quot;imports&quot;
  table for analysis. Replaces logic in the Reporter class that would,
  given an APIReport and a SDKDB instance, call getter methods one at a
  time to check for availability.
(SDKDB.audit): Runs the join query mentioned above and yields
  Diagnostics for encountered issues.
(SDKDB.symbol): Deleted.
(SDKDB.objc_class): Deleted.
(SDKDB.objc_selector): Deleted.
(SDKDB.stats): Ditto.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py:

Helpers to load fixtures, call audit(), and make assertions on the
diagnostics returned:

(TestSDKDB.add_library):
(TestSDKDB.add_allowlist):
(TestSDKDB.reconnect):
(TestSDKDB.audit_with):
(TestSDKDB.assertEmpty):

The test_audit_* functions check various cases of the audit() query.

(TestSDKDB.test_ingests_api_report_nonnormalized):
(TestSDKDB.test_entries_removed_when_binary_updated):
(TestSDKDB.test_audit_missing_name_from_spi):
(TestSDKDB.test_audit_allowed_name_from_spi):
(TestSDKDB.test_audit_api_from_loaded_file):
(TestSDKDB.test_audit_missing_name_from_unloaded_file):
(TestSDKDB.test_audit_allowed_name_from_unloaded_file):
(TestSDKDB.test_audit_unused_allow_from_api):
(TestSDKDB.test_audit_missing_name_from_unloaded_file_unloaded_allowlist):
(TestSDKDB.test_audit_allowed_own_methods):
(TestSDKDB.test_audit_unused_allow_from_loaded_allowlist):
(TestSDKDB.test_audit_no_unused_allow_from_unloaded_allowlist):
(TestSDKDB.test_audit_allowed_selector_in_loaded_and_unloaded_library):

(TestSDKDB.test_ingests_api_report): Deleted, this is the same as
  test_audit_api_from_loaded_file.
(TestSDKDB.test_only_finds_declarations_from_used_inputs): Deleted, this
  is the same as test_audit_missing_name_from_unloaded_file.

* Tools/Scripts/webkitpy/__init__.py: Bump pycodestyle to 2.14 to
  support PEP-572 := assignment expressions.

* Tools/Scripts/webkitpy/style/checker.py:
* Tools/Scripts/webkitpy/style/checkers/spi_allowlist.py: Added. Checks
  allowlist bug URLs for &quot;OOPS&quot;.
* Tools/Scripts/webkitpy/style/checkers/spi_allowlist_unittest.py: Added.

Canonical link: <a href="https://commits.webkit.org/297191@main">https://commits.webkit.org/297191@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f82a8f48e91d64c5441a147acb5ce8238108741

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110806 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20902 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116835 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112769 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39055 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/84248 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113754 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24879 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/99784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64689 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/110238 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24244 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17918 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/60630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/103298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94271 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17979 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119625 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109360 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37851 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28131 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93207 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38227 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96050 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93031 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23714 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38071 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/15829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37744 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43214 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133635 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37405 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/36091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40743 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/39112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
```diff
@@ -0,0 +1,116 @@
+from typing import Any, Optional
+
+if sys.version_info < (3, 11):
+    from pip._vendor import tomli
```
We have no guarantee that `pip` is installed, nor what version of `pip` is installed, so we absolutely cannot rely on its vendoring. If we have dependencies, we need to install them ourselves (likely via the autoinstaller).

emw-apple replied:
Hm… it's guaranteed to be present in the Xcode toolchain, is that not enough? The core `audit-spi` workflow needs to function without network access, so that precludes autoinstalling `tomli`. We could vendor tomli ourselves, but I think that's overkill when this hack _does_ work on the environments we need to support.

emw-apple replied:
One thing we could do is declare a real dependency on `tomli`, try to faithfully import it, possibly even autoinstall it, and fallback to pip's vendored copy as a last-ditch effort. That would keep this functional on internal systems with no network access, but not expose other clients to this hack. Would that be sufficient?

emw-apple replied:
Per our discussion, I've vendored tomli in `webkitapipy._vendor.tomli`.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
```diff
@@ -0,0 +1,116 @@
+    classes: list[str]
+
+
+class PermanentlyAllowedReason(Enum):
```
Maybe add a comment here saying we should just use `enum.StrEnum` once we require Python 3.11? Or even do that now, on Python 3.11 and later.

emw-apple replied:
I read the docs a bit closer, which suggest creating a StrEnum class on pre-3.11 pythons like this:

```python
class StrEnum(str, Enum):
    pass
```

which does indeed seem to work.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:
```diff
@@ -0,0 +1,88 @@
+
+class TestAllowList(TestCase):
+    def setUp(self):
+        self.tempfile = tempfile.NamedTemporaryFile(prefix='TestAllowList-')
```
We need to close this file in the `tearDown`, otherwise we'll get `ResourceWarning`.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -231,69 +294,144 @@ def add_tbd(self, tbd_file: Path,
         cur = self.con.cursor()
-        cur.execute('INSERT INTO symbol VALUES (?, ?)',
-                    (name, str(file.resolve())))
+        cur.execute(f'INSERT INTO {dest} VALUES (?, ?, ?)',
```
Can we do some validation of `dest` somehow? Otherwise this is… kinda scary?

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -231,69 +294,144 @@ def add_tbd(self, tbd_file: Path,
-                    (f'_OBJC_METACLASS_$_{name}', str(file.resolve())))
-
-    def _add_objc_selector(self, name: str, class_name: Optional[str], file: Path):
+        cur.execute(f'INSERT INTO {dest} VALUES (?, ?, ?)',
```
Can we do some validation of `dest` somehow? Otherwise this is… kinda scary?

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -231,69 +294,144 @@ def add_tbd(self, tbd_file: Path,
         cur = self.con.cursor()
-        cur.execute('INSERT INTO objc_selector VALUES (?, ?, ?)',
-                    (name, class_name, str(file.resolve())))
+        cur.execute(f'INSERT INTO {dest} VALUES (?, ?, ?)',
```
Can we do some validation of `dest` somehow? Otherwise this is… kinda scary?

briannafan commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
```diff
@@ -0,0 +1,120 @@
+                    (sels, seen_sels, 'selector', 'selectors'),
+                    (clss, seen_clss, 'class', 'classes'),
+                ):
+                    if item := entry.pop(singular, None):
```
Discussed this in-person too: It could be better to just have the allowlist formatted the same way for singular and plural key entries. It would be easier to maintain and would remove the extra logic here.

I see this being confusing for developers using these lists and an extra hurdle for them to cross (even though the code will work either way).

briannafan commented on Source/WebGPU/Configurations/AllowedSPI-legacy.toml:
```diff
@@ -0,0 +1,8 @@
+
+[legacy.legacy]
+selectors = [
+    "setSampleMask:",
```
Example of a singular entry but with plural formatting. I assume this will keep happening as more people add/remove SPIs and having one type of key format would actually be nicer than trying to be "correct" (but I appreciate the thought). And I assume no one wants to go in and clean up the legacy lists just to have consistent formatting :)

briannafan commented on Source/WebGPU/Configurations/AllowedSPI.toml:
```diff
@@ -0,0 +1,3 @@
+
+[graphics."rdar://OOPS!"]
```
Is this going to be an allowed format or is this a placeholder? If the former, could we add some warning in `program.py` that the temporary exception has no valid bug?

emw-apple replied:
Oh, yes, and this is indeed something I'd like to catch. I'll see about adding a style checking rule for this.

briannafan commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
```diff
@@ -0,0 +1,120 @@
+        seen_clss: dict[str, AllowedSPI] = {}
+        for key in doc:
+            for bug, entry in doc[key].items():
+                if bug.startswith('rdar://') or \
```
I occasionally see bugs mentioned in our code without the `https://` or as `radar://` or `rdar://problem`. Could we use some regexes here to defend against someone copy pasting a slightly different link? 

See [radar.py](https://github.com/WebKit/WebKit/blob/f8ccf55729a1d2d4fa23c98004697d869c52228b/Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py#L43) and [bugzilla.py](https://github.com/WebKit/WebKit/blob/f8ccf55729a1d2d4fa23c98004697d869c52228b/Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py#L44) for examples. 
https://github.com/WebKit/WebKit/blob/f8ccf55729a1d2d4fa23c98004697d869c52228b/Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py#L32 is more general and could be reworked for this use.

emw-apple replied:
Hm… I might actually prefer being stricter here than we are in other tools? Like, iiuc the code you linked is for detecting bug URLs in freeform text from commit logs, `git webkit` input, etc. Here, the tool can easily guide people toward using one of the standard forms, because `audit-spi` will fail to run if the bug URL provided does not match.

(see also [XKCD 1171](https://xkcd.com/1171/))

briannafan replied:
Ah, I missed that `audit-spi` would fail - that sounds good to me.

briannafan reviewed (changes requested):
Generally I'm a fan of the allowlist format and the categories will be really helpful for both adding to and reading the lists. 

And I'm sure you've noticed there's a small error in `webkitapipy.allow_unittest.TestAllowList.test_parse`.

briannafan reviewed (approved):
Thanks for the changes! LGTM after resolving the merge conflict.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
```diff
@@ -0,0 +1,116 @@
+    import tomllib
+
+if sys.version_info < (3, 11):
+    class StrEnum(str, Enum):
```
Do we not need a `__str__` here? I forget how all the magic works.

emw-apple replied:
`__str__` does not behave the same as the stdlib StrEnum, but `__format__` does, which is all that we actually need.

```python
>>> l=PermanentlyAllowedReason.LEGACY
>>> repr(l)
"<PermanentlyAllowedReason.LEGACY: 'legacy'>"
>>> str(l)
'PermanentlyAllowedReason.LEGACY'
>>> l.__format__('')
'legacy'
```

Still, I'll implement `__str__` in case it comes in handy in the future.

gsnedders reviewed (approved):
Assuming this passes EWS, r+.

emw-apple committed:
[webkitapipy] Document known SPI use via allowlists
https://bugs.webkit.org/show_bug.cgi?id=295295

Reviewed by Sam Sneddon and Brianna Fan.

Introduce a mechanism for listing allowed SPI that WebKit is using. At
build time, audit-spi reads one or more project-specific allowlists,
treats those declarations as implicitly available for SPI checking, and
ensures that the list is a subset of declarations used by the project.

Add an initial set of allowlists based on what is needed to successfully
check the current supported open source build (iOS 18.5). They are
separated into "legacy" SPI use, which was present in WebKit when iOS 18
shipped, and SPI use added since then, which is documented for cleanup.

SPI ALLOWLISTS
==============

An allowlist is a TOML document that organizes allowed declarations into
functional categories and associates each with an *exception status*
that justifies why the SPI is allowed. For example, the entry

    [graphics."rdar://147613178"]
    symbols = ["_kCAContentsFormatRGBA10XR"]

allows use of the symbol _kCAContentsFormatRGBA10XR, associates it with
the "graphics" category, and denotes a temporary exception with
rdar://147613178 as the cleanup bug.

More generically, an allowlist entry has three parts:

    [<category>.<exception>]
    symbols = [ ... ]
    classes = [ ... ]
    selectors = [ ... ]

The <category> key is an arbitrary string--its purpose is to help make
sense of why a particular SPI is being used, and to discourage
developers from simply appending to the end of the allowlist and causing
frequent merge conflicts.

The <exception> is a bug URL, denoting a temporary exception that will
be cleaned up when the bug is closed; or it's a fixed permanent
exception type. The symbols, classes, and selectors lists all correspond
to the different kinds of SPI that we track; the latter two are ObjC
specific.

Allowlists are required to be a subset of imported SPI. In other words,
it is an error for a declaration to be allowed but not used by any of
the input files given to audit-spi. This helps keep the list of
temporary exceptions accurate, and it's also a line of defense against
accidentally leaking internal-only SPI into public allowlists.

I chose TOML as the format because of its better human-readability
compared to JSON, and good language support. Python can read it
natively, so it does not preclude running audit-spi in offline build
environments where we cannot install PyYAML. Future work can adopt
TOML Kit to support audit-spi making edits to allowlists automatically,
for interactive development workflows.

IMPLEMENTATION CHANGES
======================

Restructure the SDKDB cache to have three main tables:
- "export" tracks declarations that are loaded from partial SDKDB and
  binaries, and usable as API. This merges the previous three tables.
- "allow" tracks declarations that are loaded from allow lists.
- "import" tracks declarations depended on by the input file(s) being
  analyzed. This table is ephemeral.

Instead of polling the database one declaration at a time, reimplement
the analysis as a big join query. This supports efficiently finding all
declarations which are neither allowlisted nor known to be API.
Algebraically, the join is roughly:

    ( ( import − export ) ⨯ allow ) ⨯ window

where rows that are imported but not allowed are flagged as SPI, and
rows that are allowed but not imported are flagged as needing removal.

To report results of this query, the SDKDB cache yields "diagnostic"
objects which are passed to the reporter for presentation.

* Source/WTF/Scripts/audit-spi-if-needed.sh: Pass --no-errors for now,
  as we are not yet ready to enforce the no-new-SPI workflow.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/_vendor/tomli: Added,
  vendored copy of <https://github.com/hukkin/tomli> (MIT License). This
  library is part of Python starting in 3.11, so we can remove it once
  we drop support for earlier versions.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py: Added.
(AllowedSPI):
(PermanentlyAllowedReason):
(AllowList): Data class that represents a single allowlist.
(AllowList.from_dict):
(AllowList.from_file): Main entrypoint.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
(TSVReporter.__init__): Recognize --errors and track "suggested"
  allowlists to append to when reporting unknown SPI.
(TSVReporter.emit_diagnostic): New entry point for presenting analysis
issues. Instead of querying the sdkdb cache directly, it receives
diagnostics resulting from the query in SDKDB.audit().
(TSVReporter.unused_allowed_name): Added.
(TSVReporter.finished):
(get_parser): Add a new --[no-]-errors option to control whether
potential SPI use is reported as an error.
(TSVReporter.process_report): Deleted.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(DeclarationKind): Used in the new table schema mentioned above. Rather
than have separate tables for classes/selectors/symbols, this enum is
used in a column to denote kind.
(Diagnostic): Type representing analysis issues to report. Yielded from
  audit().
(MissingName):
(UnusedAllowedName):
(SDKDB.__init__): Support marshalling DeclarationKind objects into and
  out of the database.
(SDKDB._initialize_db): Bring up new schema.
(SDKDB._initialize_temporary_schema): Ditto.
(SDKDB._add_api_report): Support adding declarations to a table other
  than the "exports" table, for tracking declarations that are imported or
  allowed rather than visible as API.
(SDKDB._add_objc_interface): Ditto.
(SDKDB._add_symbol): Ditto.
(SDKDB._add_objc_class): Ditto.
(SDKDB._add_objc_selector): Ditto.
(SDKDB.add_allowlist): Entry point for parsing an allowlist if needed,
  and updating the "allow" table. Like other input files, the allowlist
  will only be read if it's not already represented in the cache.
(SDKDB.add_for_auditing): Called to load an APIReport into the "imports"
  table for analysis. Replaces logic in the Reporter class that would,
  given an APIReport and a SDKDB instance, call getter methods one at a
  time to check for availability.
(SDKDB.audit): Runs the join query mentioned above and yields
  Diagnostics for encountered issues.
(SDKDB.symbol): Deleted.
(SDKDB.objc_class): Deleted.
(SDKDB.objc_selector): Deleted.
(SDKDB.stats): Ditto.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py:

Helpers to load fixtures, call audit(), and make assertions on the
diagnostics returned:

(TestSDKDB.add_library):
(TestSDKDB.add_allowlist):
(TestSDKDB.reconnect):
(TestSDKDB.audit_with):
(TestSDKDB.assertEmpty):

The test_audit_* functions check various cases of the audit() query.

(TestSDKDB.test_ingests_api_report_nonnormalized):
(TestSDKDB.test_entries_removed_when_binary_updated):
(TestSDKDB.test_audit_missing_name_from_spi):
(TestSDKDB.test_audit_allowed_name_from_spi):
(TestSDKDB.test_audit_api_from_loaded_file):
(TestSDKDB.test_audit_missing_name_from_unloaded_file):
(TestSDKDB.test_audit_allowed_name_from_unloaded_file):
(TestSDKDB.test_audit_unused_allow_from_api):
(TestSDKDB.test_audit_missing_name_from_unloaded_file_unloaded_allowlist):
(TestSDKDB.test_audit_allowed_own_methods):
(TestSDKDB.test_audit_unused_allow_from_loaded_allowlist):
(TestSDKDB.test_audit_no_unused_allow_from_unloaded_allowlist):
(TestSDKDB.test_audit_allowed_selector_in_loaded_and_unloaded_library):

(TestSDKDB.test_ingests_api_report): Deleted, this is the same as
  test_audit_api_from_loaded_file.
(TestSDKDB.test_only_finds_declarations_from_used_inputs): Deleted, this
  is the same as test_audit_missing_name_from_unloaded_file.

* Tools/Scripts/webkitpy/__init__.py: Bump pycodestyle to 2.14 to
  support PEP-572 := assignment expressions.

* Tools/Scripts/webkitpy/style/checker.py:
* Tools/Scripts/webkitpy/style/checkers/spi_allowlist.py: Added. Checks
  allowlist bug URLs for "OOPS".
* Tools/Scripts/webkitpy/style/checkers/spi_allowlist_unittest.py: Added.

Canonical link: https://commits.webkit.org/297191@main
